src/libxtp/gaussian_quadrature.cc

```diff
@@ -0,0 +1,225 @@
+/*
+ *            Copyright 2009-2019 The VOTCA Development Team
+ *                       (http://www.votca.org)
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License")
+ *
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *              http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <votca/tools/constants.h>
+#include <votca/xtp/gauss_hermite_quadrature_constants.h>
+#include <votca/xtp/gauss_laguerre_quadrature_constants.h>
+#include <votca/xtp/gauss_legendre_quadrature_constants.h>
+#include <votca/xtp/gaussian_quadrature.h>
+#include <votca/xtp/threecenter.h>
+
+namespace votca {
+namespace xtp {
+
+// Constructor
+GaussianQuadrature::GaussianQuadrature(const Eigen::VectorXd& energies,
+                                       const TCMatrix_gwbse& Mmn)
+    : _energies(energies), _Mmn(Mmn) {}
+
+void GaussianQuadrature::configure(options opt, const RPA& rpa) {
+  _opt = opt;
+  if (_opt.quadrature_scheme == "laguerre") {
+    Gauss_Laguerre_Quadrature_Constants glqc;
+    _quadpoints = glqc.getPoints(_opt.order);
+    _quadadaptedweights = glqc.getAdaptedWeights(_opt.order);
+    CalcDielInvVector(rpa);
+  } else if (_opt.quadrature_scheme == "legendre") {
+    Gauss_Legendre_Quadrature_Constants glqc;
+    _quadpoints = glqc.getPoints(_opt.order);
+    _quadadaptedweights = glqc.getAdaptedWeights(_opt.order);
+    CalcDielInvVector(rpa);
+  } else if (_opt.quadrature_scheme == "modified_legendre") {
+    Gauss_Legendre_Quadrature_Constants glqc;
+    _quadpoints = glqc.getPoints(_opt.order);
+    _quadadaptedweights = glqc.getAdaptedWeights(_opt.order);
+    CalcDielInvVector(rpa);
+  } else if (_opt.quadrature_scheme == "hermite") {
+    Gauss_Hermite_Quadrature_Constants glqc;
+    _quadpoints = glqc.getPoints(_opt.order);
+    _quadadaptedweights = glqc.getAdaptedWeights(_opt.order);
+    CalcDielInvVector(rpa);
+  } else {
+    std::cout << "There no such a thing as the integration scheme you asked"
+              << std::endl;
+  }
+}
+
+// This function calculates and stores inverses of the microscopic dielectric
+// matrix in a matrix vector
+void GaussianQuadrature::CalcDielInvVector(const RPA& rpa) {
+  _dielinv_matrices_r.resize(_opt.order);
+  double halfpi = std::acos(0.0);
+#pragma openmp parallel schedule(guided)
```

Comment:
This only parallelizes up to the order which can be small. Is that the parallelisation scheme you want?

Reply:
I inherited this from another implementation so I am not fond of it. What parallel scheme do you suggest to use?

Reply:
I have no idea at the moment. If you want a more fine-grained parallelism you probably have to go one level deeper in the code.

Reply:
I am not sure how critical it is, how important this part is, but if `order=12` then you can only use upt to 12 threads to parallelize it.

Reply:
Ok maybe we should go for a static schedule, right?

Reply:
It will not really help either. If you have more threads than `order` on your machine, the remaining threads will idle.

Reply:
There is also parallelization in `rpa.calculate_epsilon_complex()`, so what happens if we have a parallel region inside a parallel region? Also, is the `inverse()` call in eigen using a parallel MKL backend?

Reply:
That is a good question. So I think in general nested parallelism is enabled. So we have a pool of threads, whose size is determined by `omp_set_num_thread`, if the first `omp_parallel_for` does not require all threads, it will use those threads further down the line.

Reply:
Same is true for MKL, as long as it uses OPENMP for threading, which it does in votca.

Reply:
So it is pretty unlikely that having parallelization here is going to have much of an effect, because probably the parallel sum over occ states in RPA is more relevant anyways, I assume.

Reply:
yes